### ISSUE TITLE
Asset files not automatically uploading to S3 bucket

### ISSUE THREAD

**Issue by funkel1989** [wan2land/serverless-nuxt#311]
Hello, I attempted to follow the instructions here and deploy a site using this but I'm having an issue with any of my static assets loading. it seems only the favicon is loading on the site and not of the images I've placed. It also seems like some of my styling is not correctly loaded but everything works fine in DEV. My assumption is I'm doing something wrong with the CDN. Can anyone take a look and let me know if I've done this correctly?
https://github.com/Battle-Line-Productions/battlesite-home

**Comment by funkel1989** [comment]
So after doing a git of research into this I've noticed that none of the nuxt generated client assets are being uploaded to the  S3 bucket. I'm assuming the serverless plugin is supposed to handle this but I've seen no reference to it when i read the docs. I also don't see any reference to where static files are being uploaded for the lambda to reference, if that's part of the assets or if I need to do something extra for static images and what not. Any help here would be really appreciated.

**Comment by funkel1989** [comment]
So just one last update for tonight. Turns out that network tab of chrome is throwing a 403 error when trying to access the assets that are on the asset bucket (which i still had to manually upload). super confused here and could use some help understanding thats happening.

**Comment by funkel1989** [comment]
So i was able to get the site to work, and frankly i'm still not sure what I did to make it work right but when I can say is the assets are still not uploading to the Assets bucket during the deploy, I've had to manually do that. it looks like the code in your plugin should be handling the upload but i'm also not getting any errors. Posting some code files below:

```yml
service: battle-site-home
frameworkVersion: ^2.35.0"

provider:
  name: aws
  runtime: nodejs12.x
  region: us-east-1
  stage: ${opt:stage, 'dev'}
  profile: ${opt:profile, 'blptest'}
  lambdaHashingVersion: '20201221'
  environment:
    NODE_ENV: ${env:NODE_ENV}
  iamRoleStatements:
    - Effect: 'Allow'
      Action:
        - 's3: *'
      Resource:
        - 'arn:aws:s3:::${self:custom.nuxt.bucketName}/*'

plugins:
  - serverless-nuxt-plugin
  - serverless-apigw-binary
  - serverless-domain-manager

custom:
  customDomain:
    domainName: ${env:DOMAIN_NAME}
  apigwBinary:
    types:
      - 'image/*'
  nuxt:
    version: v${file(./package.json):version}
    bucketName: battle-line-productions-${self:provider.stage}-v2
    cdnPath: https://${env:ASSETS_DOMAIN_NAME}

resources:
  Resources:
    AssetsBucket:
      Type: AWS::S3::Bucket
      Properties:
        BucketName: ${self:custom.nuxt.bucketName}
        AccessControl: Private
        CorsConfiguration:
          CorsRules:
            - AllowedMethods:
              - GET
              - HEAD
              AllowedOrigins:
              - https://${env:DOMAIN_NAME}
    AssetsUIBucketPolicy:
      Type: AWS::S3::BucketPolicy
      Properties:
        Bucket:
          Ref: AssetsBucket
        PolicyDocument:
          Statement:
            - Action:
                - 's3:GetObject'
              Effect: Allow
              Resource:
                {
                  'Fn::Join':
                    ['', ['arn:aws:s3:::', { Ref: AssetsBucket }, '/*']],
                }
              Principal:
                CanonicalUser:
                  'Fn::GetAtt':
                    [AssetsUIOriginAccessIdentity, S3CanonicalUserId]
    # RedirectBucket:
    #   Type: AWS::S3::Bucket
    #   Properties:
    #     BucketName: ${env:REDIRECT_DOMAIN_NAME}
    #     WebsiteConfiguration:
    #       RedirectAllRequestsTo:
    #         HostName: ${env:DOMAIN_NAME}
    #         Protocol: https
    # RedirectRoute53ARecord:
    #   Type: AWS::Route53::RecordSet
    #   Properties:
    #     HostedZoneName: ${env:DOMAIN_HOSTED_ZONE}
    #     Name: ${env:REDIRECT_DOMAIN_NAME}
    #     Type: A
    #     AliasTarget:
    #       DNSName: s3-website.us-east-2.amazonaws.com
    #       HostedZoneId: Z075584225FDBNT9EDT7R
    #       EvaluateTargetHealth: false
    AssetsUIOriginAccessIdentity:
      Type: AWS::CloudFront::CloudFrontOriginAccessIdentity
      Properties:
        CloudFrontOriginAccessIdentityConfig:
          Comment: 'CloudFrontOriginAccessIdentity for ${self:service}-${self:provider.stage}'
    AssetsCloudFrontDistribution:
      Type: AWS::CloudFront::Distribution
      Properties:
        DistributionConfig:
          Enabled: true
          HttpVersion: http2
          PriceClass: PriceClass_100
          Restrictions:
            GeoRestriction:
              RestrictionType: whitelist
              Locations:
                - US
          Aliases:
            - ${env:ASSETS_DOMAIN_NAME}
          Origins:
            - Id: AssetsBucket
              DomainName:
                'Fn::GetAtt': [AssetsBucket, DomainName]
              S3OriginConfig:
                OriginAccessIdentity:
                  'Fn::Join':
                    [
                      '',
                      [
                        'origin-access-identity/cloudfront/',
                        { Ref: AssetsUIOriginAccessIdentity },
                      ],
                    ]
          DefaultCacheBehavior:
            AllowedMethods: [DELETE, GET, HEAD, OPTIONS, PATCH, POST, PUT]
            CachedMethods: [GET, HEAD, OPTIONS]
            TargetOriginId: AssetsBucket
            DefaultTTL: 0
            MaxTTL: 0
            MinTTL: 0
            Compress: true
            ForwardedValues:
              QueryString: false
              Cookies:
                Forward: all
            ViewerProtocolPolicy: redirect-to-https
          ViewerCertificate:
            AcmCertificateArn: ${env:ASSETS_ACM_ARN}
            SslSupportMethod: sni-only
            MinimumProtocolVersion: TLSv1.1_2016
    AssetsRoute53ARecord:
      Type: AWS::Route53::RecordSet
      Properties:
        HostedZoneName: ${env:DOMAIN_HOSTED_ZONE}
        Name: ${env:ASSETS_DOMAIN_NAME}
        Type: A
        AliasTarget:
          DNSName: !GetAtt AssetsCloudFrontDistribution.DomainName
          HostedZoneId: Z2FDTNDATAQYW2
          EvaluateTargetHealth: false
    AssetsRoute53AAAARecord:
      Type: AWS::Route53::RecordSet
      Properties:
        HostedZoneName: ${env:DOMAIN_HOSTED_ZONE}
        Name: ${env:ASSETS_DOMAIN_NAME}
        Type: AAAA
        AliasTarget:
          DNSName: !GetAtt AssetsCloudFrontDistribution.DomainName
          HostedZoneId: Z2FDTNDATAQYW2
          EvaluateTargetHealth: false

functions:
  nuxt:
    memorySize: 2048
    timeout: 30
    handler: handler.render
    events:
      - http: ANY /
      - http: ANY /{proxy+}


Outputs:
  AssetsCloudFrontDistributionId:
    Value:
      Ref: AssetsCloudFrontDistribution
  AssetsCloudFrontDistributionDomainName:
    Value:
      'Fn::GetAtt': [AssetsCloudFrontDistribution, DomainName]

```


```javascript
const config = require('./src/configs');

const { gaId } = config.analytics

module.exports = {
  // ssr: false,
  target: 'server',
  telemetry: false,
  srcDir: 'src/',
  // Global page headers (https://go.nuxtjs.dev/config-head)
  head: {
    titleTemplate: '%s',
    title: 'Battleline Productions',
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'description', name: 'description', content: '' }
    ],
    link: [
      { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' },
      { rel: 'stylesheet', href: 'https://fonts.googleapis.com/css2?family=Heebo:wght@100;200;300;400;600;700;800;900&display=swap' },
      ...config.icons.map((href) => ({ rel: 'stylesheet', href }))
    ]
  },

  // Global CSS (https://go.nuxtjs.dev/config-css)
  css: [
    '~/assets/scss/theme.scss'
  ],

  // Plugins to run before rendering page (https://go.nuxtjs.dev/config-plugins)
  plugins: [
    // filters
    { src: '~/filters/capitalize.js' },
    { src: '~/filters/lowercase.js' },
    { src: '~/filters/uppercase.js' },
    { src: '~/filters/trim.js' },
    { src: '~/filters/placeholder.js' },
    { src: '~/filters/formatDate.js' }
  ],

  // Auto import components (https://go.nuxtjs.dev/config-components)
  // components: true,

  // Modules for dev and build (recommended) (https://go.nuxtjs.dev/config-modules)
  buildModules: [
    // https://go.nuxtjs.dev/vuetify
    ['@nuxtjs/vuetify', {
      customVariables: ['~/assets/scss/vuetify/variables/_index.scss'],
      optionsPath: '~/configs/vuetify.js',
      treeShake: true,
      defaultAssets: {
        font: false
      }
    }]
  ],

  // Modules (https://go.nuxtjs.dev/config-modules)
  modules: [
    '@nuxtjs/google-gtag'
  ],

  'google-gtag': {
    id: gaId,
    debug: true, // enable to track in dev mode
    disableAutoPageTrack: false // disable if you don't want to track each page route with router.afterEach(...).
  },

  // Build Configuration (https://go.nuxtjs.dev/config-build)
  build: {
    publicPath: process.env.SERVERLESS_NUXT_PUBLIC_PATH
  }
}
```

```json
{
  "name": "@battle-line-productions/battlesite-home",
  "version": "1.2.0",
  "private": true,
  "scripts": {
    "dev": "nuxt",
    "build": "nuxt build",
    "start": "nuxt start",
    "generate": "nuxt generate",
    "domain": "npx env-cmd -e prod sls create_domain",
    "undomain": "npx env-cmd -e prod sls remove_domain",
    "deploy": "npx env-cmd -e prod sls deploy --force --stage prod --verbose",
    "remove": "sls remove --stage prod"
  },
  "dependencies": {
    "@nuxtjs/google-gtag": "^1.0.4",
    "@nuxtjs/vuetify": "^1.11.3",
    "core-js": "^3.10.1",
    "date-fns": "^2.20.1",
    "lodash": "^4.17.21",
    "nuxt-start": "^2.15.4",
    "serverless-nuxt": "^1.1.1"
  },
  "devDependencies": {
    "@indielayer/eslint-config-vue": "^2.1.2",
    "env-cmd": "^10.1.0",
    "eslint": "^7.24.0",
    "nuxt": "^2.15.4",
    "sass": "^1.32.8",
    "sass-loader": "10.1.1",
    "serverless": "^2.35.0",
    "serverless-apigw-binary": "^0.4.4",
    "serverless-domain-manager": "^5.1.0",
    "serverless-nuxt-plugin": "^1.3.0"
  }
}

```

**Comment by wan2land** [comment]
In Nuxt, files uploaded to `static` are handled within the same domain. It is difficult to handle in Lambda.

So, move all `static` files except `favicon` to `assets` directory.

in css,

```css
div {
  background: url(~/assets/images/something.png)
}
```

in html,

```
<template>
  <div><img :src="require('~/assets/images/something.png')" /></div>
</template>
```

**Comment by funkel1989** [comment]
@wan2land That is very helpful to know. Thank you very much but I think the issue i'm having might have been misunderstood. At this current point in time the problem is, after a serverless deploy completes the asset bucket is empty. No assets at all are being uploaded.

The plugin is executing the code to upload the client asset files and shows logs during the deploy like it is working but nothing ever gets uploaded.

I'm talking about the assets from /_nuxt/dist/client/   << i've tried defining this as the cdnPath even though it should be the default and it doesn't seem to change anything.

**Comment by funkel1989** [comment]
@wan2land I guess I'll take the eyes as you've never seen this before and don't really have any idea. I've dug a little bit and from what I can tell the asset path is correct all the way through your code. globby is not returning any assetFiles and instead returns an empty array which is the problem. I tried removing the only files pieces but that made no different.

My asset path ends up looking like this "C:\\GIT\\BattleLineProductions\\nuxt\\.nuxt\\dist\\client" which is valid. nuxt is the project folder and at build time it contains a .nuxt dir exactly as expected.

**Comment by funkel1989** [comment]
@wan2land So digging a bit deeper today it looks like this is a known issue where globby simply isn't working on windows machines. It also kind of sounds like they don't have plans to fix it anytime soon. You might want to consider an alternative to globby for the future of this plugin if you want to support both linux, mac and windows machines. In the mean time i've manually edited your plugin file as below to make this work for me.

```javascript
this.serverless.cli.consoleLog(`Serverless Nuxt Plugin: ${chalk.yellow('upload asset files')}`)
    const assetsFiles = await globby(assetsPath.replace(/\\/g, '//'), { onlyFiles: true, expandDirectories: { files: ['*.*'], extensions: ['*'] } })
    console.log(JSON.stringify(assetsFiles, null, 2));
    await Promise.all(assetsFiles.map((file) => {
      const fileTargetPath = [config.version, file.replace(assetsPath.replace(/\\/g, '//'), '')].join('/')
        .replace(/^\/+|\/+$/, '')
        .replace(/\/+/g, '/')
      console.log(`Putting file: ${fileTargetPath}`);
      return s3.putObject({
        Bucket: config.bucketName,
        Key: fileTargetPath,
        Body: fs.readFileSync(file),
        ACL: 'public-read',
        ContentType: mime.lookup(file) || null,
        CacheControl: `public, max-age=${Math.floor(ms(config.assetsCacheMaxAge) / 1000)}`, // let the browser cache all static files for 1 year since they receive a unique hash by Nuxt
      }).promise().catch(ex => console.log(ex))
    }))
```

**Comment by funkel1989** [comment]
Nevermind, the above. I'm leaving the comment there because it does fix globby but then the assetPath's and keys uploaded to the S3 bucket are a wreck. I ended up uploading the paths all the way back from my C drive. This needs some work on windows it seems.

**Comment by wan2land** [comment]
Does the above method work in a Windows environment?  
As soon as the Windows environment is ready, I will refer to the above source and modify it.  
thank you.

**Comment by funkel1989** [comment]
@wan2land the above referenced code fixes finding the files to upload in a windows environment but will break Linux/mac (i think?). I've found another issue though that when the files are uploaded to the S3 bucket they include the full path of the file at the file key.

Example  nuxt/.nuxt/dist/client/file.jpg which i expect to upload to the s3 bucket as {{version}}/file.jpg.

instead this ends up in the s3 bucket as C://git/org/nuxt/.nuxt/disk/client/{{version}}/file.jpg

The code difference between the different OS's seems to be decent. 

I ended up giving up and solving my problem by setting up a deploy on commit using Ubuntu running on GitHub actions (finally got this working also!).
